vmi ctrl: Decouple network info annotation gen

orelmisan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Currently, the logic responsible for calculating the network info annotation is coupled with the VMI controller
and its tests.

Introduce the `annotationsGenerator` interface in the VMI controller.

Move the concrete logic from `pkg/virt-controller/network` package to `pkg/network/pod/annotations`
in order to have all pod annotations related code under the same package.

Following this change, the logic responsible for Multus network annotation update due to hotplug/hotunplug
is still present in the VMI controller, and will be extracted by a follow-up PR.

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Fixes #

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [X] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
None
```

orelmisan commented:
/uncc @AlonaKaplan @alicefr 
/cc @EdDev

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/15f735d0f7aaf24ad57ef1d12c803ef69e872041..a3196cb70e4da7cc556cba0db1e5763a28c2205f): Removed additional unused functions from `vmi_test.go`.

Acedus commented on pkg/virt-controller/watch/vmi/vmi.go:
```diff
@@ -171,25 +173,30 @@ func (i informalSyncError) RequiresRequeue() bool {
 	return false
 }
 
+type annotationsGenerator interface {
```
While I understand this is done in order to simplify vmi_test by stubbing, what do you think about introducing the interface as part of network/pod/annotations instead? The only annotation generator that actually has the `GenerateFromPod` method is the network one, I think that changing the type to `<package>.annotationGenerator` would better provide the necessary context and allow future extensibility.

Acedus replied:
I retract this comment, after discussing it we've concluded that it may serve future generators.

orelmisan replied:
This is done in order to decouple the concrete logic from the VMI controller.
As a result of using this interface, we are able to use a stub in the VMI controller's unit tests.

In Go, interfaces are implemented implicitly, which mean we can describe the behavior a struct/function/method expects from its input.
This is the reason the interface in question is present in this specific location.

Exported interfaces should be interfaces which are very common and useful, for example [io.Reader](https://pkg.go.dev/io#Reader).

I assumed this interface could be beneficial to other SIGs, so I haven't prefixed it with `net`.

Acedus reviewed (commented):
Thank you, please see inline comments.

EdDev commented on pkg/network/pod/annotations/generator.go:
```diff
@@ -98,6 +102,37 @@ func (g Generator) GenerateFromSource(vmi *v1.VirtualMachineInstance, sourcePod
+	annotations := map[string]string{}
+
+	netInfoAnnotation, err := g.generateNetworkInfoAnnotation(vmi, pod)
+	if err != nil {
+		return nil, err
+	}
```
This is a logic change which has been introduced without explanation.
The logic has ignored such errors and just logged them, resulting on the whole reconcile not to fail but to just skip the network info update.

Could you please reason on why this change needs to do this?

EdDev commented on pkg/network/pod/annotations/generator.go:
```diff
@@ -98,6 +102,37 @@ func (g Generator) GenerateFromSource(vmi *v1.VirtualMachineInstance, sourcePod
 	return annotations, nil
 }
 
+func (g Generator) GenerateFromPod(vmi *v1.VirtualMachineInstance, pod *k8Scorev1.Pod) (map[string]string, error) {
```
Under this file, we have two similar methods: `GenerateFromSource` and `GenerateFromPod`.
The name is also confusing: it mentions "from pod", but it is "from vmi and pod".

Perhaps `GenerateFromActive` or `GenerateFromTarget` is a better option.
A description for the two will also help, clarifying that is the expected pod input.

orelmisan replied:
Renamed `GenerateFromPod` to `GenerateFromActivePod`.

orelmisan commented:
Change: Answered @EdDev's comments.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/a7e6d859e47a3a4d0c959259969018a363b7e211..ac82e0b83443248aaf8a87f67330a0cbef1bb534): Reverted some changes made to the original implementation.

orelmisan commented on pkg/virt-controller/watch/vmi/vmi_test.go:
```diff
@@ -742,80 +743,62 @@ var _ = Describe("VirtualMachineInstance watcher", func() {
 		)
 
-		It("should not patch network-info annotation when no SR-IOV/binding plugin networks exist", func() {
+		It("should patch network annotations", func() {
```
Done.

EdDev commented on pkg/virt-controller/watch/vmi/vmi_test.go:
```diff
@@ -742,80 +742,44 @@ var _ = Describe("VirtualMachineInstance watcher", func() {
 		})
 
-		It("should patch network-info annotation when SR-IOV networks exist", func() {
+		It("should not patch the pod's annotations when network annotations generator returns an empty map", func() {
```
Can you set these annotations explicitly empty? It is now implicitly empty, which is hard to follow.

orelmisan replied:
Done.

EdDev commented on pkg/network/pod/annotations/generator_test.go:
```diff
@@ -247,6 +250,166 @@ var _ = Describe("Annotations Generator", func() {
+			nonDeviceInfoPlugin = "non_deviceinfo"
+		)
+
+		const (
+			multusNetworkStatusWithPrimaryNet = `[{"name":"k8s-pod-network","ips":["10.244.196.146","fd10:244::c491"],"default":true,"dns":{}}]`
+
+			multusNetworkStatusWithPrimaryAndSecondaryNetsWithoutDeviceInfo = `[{"name":"k8s-pod-network","ips":["10.244.196.146","fd10:244::c491"],"default":true,"dns":{}},
+{"name":"default/no-device-info","interface":"pod6446d58d6df","mac":"8a:37:d9:e7:0f:18","dns":{}}]`
+
+			multusNetworkStatusWithPrimaryAndSecondaryNetsWithDeviceInfo = `[{"name":"k8s-pod-network","ips":["10.244.196.146","fd10:244::c491"],"default":true,"dns":{}},
+{"name":"default/with-device-info","interface":"pod2c26b46b68f","dns":{},"device-info":{"type":"pci","version":"1.0.0","pci":{"pci-address":"0000:65:00.2"}}}]`
+
+			multusNetworkStatusWithPrimaryAndSRIOVSecondaryNet = `[{"name":"k8s-pod-network","ips":["10.244.196.146","fd10:244::c491"],"default":true,"dns":{}},
+{"name":"default/sriov","interface":"pod778c553efa0","dns":{},"device-info":{"type":"pci","version":"1.0.0","pci":{"pci-address":"0000:65:00.3"}}}]`
+
+			multusNetworkStatusWithPrimaryAndFourSecondaryNets = `[{"name":"k8s-pod-network","ips":["10.244.196.146","fd10:244::c491"],"default":true,"dns":{}},
+{"name":"default/with-device-info","interface":"pod2c26b46b68f","dns":{},"device-info":{"type":"pci","version":"1.0.0","pci":{"pci-address":"0000:65:00.2"}}},
+{"name":"default/sriov","interface":"pod778c553efa0","dns":{},"device-info":{"type":"pci","version":"1.0.0","pci":{"pci-address":"0000:65:00.3"}}},
+{"name":"default/no-device-info","interface":"pod6446d58d6df","mac":"8a:37:d9:e7:0f:18","dns":{}},
+{"name":"default/br-net","interface":"podeeea394806a","mac":"6a:1f:28:23:58:40","dns":{}}]`
+		)
```
PLease move the consts closer to the usage, it is easier to read the test that way.
I think there is only constant that is used twice, I would just define it in each location.

orelmisan replied:
Done.

EdDev reviewed (commented):
Thanks.

Please try to cover `network/pod` with the linter. I have missed that when it was added.

orelmisan commented:
> Thanks.
> 
> Please try to cover `network/pod` with the linter. I have missed that when it was added.

Please see https://github.com/kubevirt/kubevirt/pull/12846.

EdDev commented:
> > Thanks.
> > Please try to cover `network/pod` with the linter. I have missed that when it was added.
> 
> Please see #12846.

You should cherry-pick it here to show it is also good. Whichever enters first, wins.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/a3d9cf3392c9f2041fc67d7ddb1d731b042cdcdc..a9ff58785442c4982c0cb1b040f6792473b3e734): Rebased on top of #12846.

[Change](https://github.com/kubevirt/kubevirt/compare/a9ff58785442c4982c0cb1b040f6792473b3e734..b66fe83c16ffa60a43934680f3af3e3e547618ea): Dropped the storage related commit.

orelmisan commented:
> > > Thanks.
> > > Please try to cover `network/pod` with the linter. I have missed that when it was added.
> > 
> > 
> > Please see #12846.
> 
> You should cherry-pick it here to show it is also good. Whichever enters first, wins.

Done.

EdDev reviewed (approved):
Thank you!

/approve

We need help with approving the non-network areas.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/b66fe83c16ffa60a43934680f3af3e3e547618ea..22f74244ebf416bb077f2d6b73c1e71c1508834b): Rebase

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/22f74244ebf416bb077f2d6b73c1e71c1508834b..f442fcc08f85e3c284623ccbf918493154d7e059): Drop `nolint` directives and make the linter happy.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/f442fcc08f85e3c284623ccbf918493154d7e059..ba3b5cc057cc42af7e684908e7dafe7789f3d405): Align `multusNetworkStatusEntryForDeviceInfo` with `multusNetworkStatusEntryForSRIOV`.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/ba3b5cc057cc42af7e684908e7dafe7789f3d405..3085522bd977fcfe4aea90b59a2a2960a7b3a6ac): Rebase

EdDev reviewed (approved):
Thanks!

/approve

enp0s3 reviewed (commented):
/approve
